src/comm/XPUPair.h

```diff
@@ -0,0 +1,246 @@
+#pragma once
+
+#include <ATen/Context.h>
+
+namespace at::xpu {
+
+// define null_type
+struct null_type {};
+
+// null_type comparisons
+inline bool operator==(const null_type&, const null_type&) {
+  return true;
+}
+
+inline bool operator>=(const null_type&, const null_type&) {
+  return true;
+}
+
+inline bool operator<=(const null_type&, const null_type&) {
+  return true;
+}
+
+inline bool operator!=(const null_type&, const null_type&) {
+  return false;
+}
+
+inline bool operator<(const null_type&, const null_type&) {
+  return false;
+}
+
+inline bool operator>(const null_type&, const null_type&) {
+  return false;
+}
+
+// forward declaration for tuple
+template <
+    class T0 = null_type,
+    class T1 = null_type,
+    class T2 = null_type,
+    class T3 = null_type,
+    class T4 = null_type,
+    class T5 = null_type,
+    class T6 = null_type,
+    class T7 = null_type,
+    class T8 = null_type,
+    class T9 = null_type>
+class tuple;
+
+template <typename T1, typename T2>
+inline void swap(T1& a, T2& b) {
+  T1 temp = a;
+  a = b;
+  b = temp;
+}
+
+template <typename T1, typename T2>
+struct pair {
+  typedef T1 first_type;
+  typedef T2 second_type;
+
+  first_type first;
+  second_type second;
+
+  // default constructor
+  pair(void) : first(), second() {}
+
+  inline pair(const T1& x, const T2& y) : first(x), second(y) {}
+
+  template <typename U1, typename U2>
+  inline pair(const pair<U1, U2>& p) : first(p.first), second(p.second) {}
+
+  template <typename U1, typename U2>
+  pair(const std::pair<U1, U2>& p) : first(p.first), second(p.second) {}
+
+  inline void swap(pair& p) {
+    swap(first, p.first);
+    swap(second, p, second);
+  }
```

Comment:
Should use `std::swap` and typo "p, second".

```suggestion
  inline void swap(pair& p) {
    using std::swap;
    swap(first, p.first);
    swap(second, p.second);
  }
```